File: Ruby/lib/civitas-ruby/mazo_sorpresas.rb
module Civitas
  class MazoSorpresas
    
    def initialize(debug = false)
      @sorpresas = []
      @barajada = false
      @usadas = 0
      @debug = debug
      Diario.instance.ocurre_evento("MazoSorpresas puesto en modo debug") if debug
      @cartas_especiales = []
      @ultima_sorpresa
    end

    def al_mazo(s)
      @sorpresas << s if !@barajada
    end

    def siguiente
      if !@barajada || (@usadas == @sorpresas.size)
        @sorpresas.shuffle! if !@debug
        @usadas = 0
        @barajada = true
      end
      @usadas += 1
      @ultima_sorpresa = @sorpresas.shift
      @sorpresas << @ultima_sorpresa
      @ultima_sorpresa
    end

    def inhabilitar_carta_especial(sorpresa)
      inhabilitada = @sorpresas.delete(sorpresa)
      if inhabilitada != nil
        @cartas_especiales << inhabilitada
        Diario.instance.ocurre_evento("Inhabilitada sorpresa #{sorpresa}")
      end
    end
    
    def habilitar_carta_especial(sorpresa)
      habilitada = @cartas_especiales.delete(sorpresa)
      if habilitada != nil
        @sorpresas << habilitada
        Diario.instance.ocurre_evento("Habilitada sorpresa #{sorpresa}")
      end
    end

  end
end

File: civitas/MazoSorpresas.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;

public class MazoSorpresas {

  private Boolean barajada;
  private Boolean debug;
  private Sorpresa ultimaSorpresa;
  private int usadas;
  private ArrayList<Sorpresa> sorpresas;
  private ArrayList<Sorpresa> cartasEspeciales;

  void alMazo (Sorpresa s) {
    if (!barajada) {
      sorpresas.add(s);
    }
  }

  Sorpresa getUltimaSorpresa() {
    return ultimaSorpresa;
  }

  void habilitarCartaEspecial (Sorpresa sorpresa) {
    int ubicEspecial = cartasEspeciales.indexOf(sorpresa);
    if (ubicEspecial > -1) {
      Sorpresa aDevolver = cartasEspeciales.remove(ubicEspecial);
      sorpresas.add(aDevolver);
      Diario.getInstance().ocurreEvento("Se devuelve una sopresa al mazo");
    }
  }
  
  void inhabilitarCartaEspecial (Sorpresa sorpresa) {
    int ubicSorpresa = sorpresas.indexOf(sorpresa);
    if (ubicSorpresa > -1) {
      Sorpresa aRemover = sorpresas.remove(ubicSorpresa);
      cartasEspeciales.add(aRemover);
      Diario.getInstance().ocurreEvento("Se añade una sorpresa a cartasEspeciales");
    }
  }

  private void init () {
    sorpresas = new ArrayList<>();
    cartasEspeciales = new ArrayList<>();
    barajada = false;
    usadas = 0;
  }

  MazoSorpresas (Boolean b) {
    debug = b;
    init();
    if (debug) {
      Diario.getInstance().ocurreEvento("Debug = True");
    }
  }

  MazoSorpresas () {
      debug = false;
      init();
  }

  Sorpresa siguiente () {
    if (!barajada || usadas == sorpresas.size()) {
      if (!debug) {
        Collections.shuffle(sorpresas);
        usadas = 0;
        barajada = true;
      }
    }
    usadas++;
    ultimaSorpresa = sorpresas.remove(0);
    sorpresas.add(ultimaSorpresa);
    return ultimaSorpresa;
  }

}
